controllers/sandbox_controller.go

```diff
@@ -70,76 +78,143 @@ func (r *SandboxReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
-		if err = r.Create(ctx, pod, client.FieldOwner("sandbox-controller")); err != nil {
-			log.Error(err, "Failed to create", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-			return ctrl.Result{}, err
+
+	// Update the sandbox status
+	// TODO: Update the sandbox status
+
+	return ctrl.Result{}, nil
+}
+
+// NameHash generates an FNV-1a hash from a string and returns
+// it as a fixed-length hexadecimal string.
+func NameHash(objectName string) string {
+	h := fnv.New32a()
+	h.Write([]byte(objectName))
+	hashValue := h.Sum32()
+
+	// Convert the uint32 to a hexadecimal string.
+	// This results in an 8-character string (e.g., "a5b3c2d1").
+	return fmt.Sprintf("%08x", hashValue)
+}
+
+func (r *SandboxReconciler) reconcileService(ctx context.Context, sandbox *sandboxv1alpha1.Sandbox, nameHash string) (*corev1.Service, error) {
+	log := log.FromContext(ctx)
+	service := &corev1.Service{}
+	found := false
+	if err := r.Get(ctx, types.NamespacedName{Name: sandbox.Name, Namespace: sandbox.Namespace}, service); err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "Failed to get Service")
+			return nil, err
 		}
 	} else {
+		found = true
+	}
+
+	if found {
+		log.Info("Found Service", "Service.Namespace", service.Namespace, "Service.Name", service.Name)
+		return service, nil
+	}
+
+	log.Info("Creating a new Headless Service", "Service.Namespace", sandbox.Namespace, "Service.Name", sandbox.Name)
+	service = &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sandbox.Name,
+			Namespace: sandbox.Namespace,
+			Labels: map[string]string{
+				sandboxLabel: nameHash,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			ClusterIP: "None",
+			Selector: map[string]string{
+				sandboxLabel: nameHash,
+			},
+		},
+	}
+	service.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Service"))
+	if err := ctrl.SetControllerReference(sandbox, service, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	if err := r.Create(ctx, service, client.FieldOwner("sandbox-controller")); err != nil {
+		log.Error(err, "Failed to create", "Service.Namespace", service.Namespace, "Service.Name", service.Name)
+		return nil, err
+	}
+	return service, nil
+}
+
+func (r *SandboxReconciler) reconcilePod(ctx context.Context, sandbox *sandboxv1alpha1.Sandbox, nameHash string) (*corev1.Pod, error) {
+	log := log.FromContext(ctx)
+	found := false
+	pod := &corev1.Pod{}
+	if err := r.Get(ctx, types.NamespacedName{Name: sandbox.Name, Namespace: sandbox.Namespace}, pod); err != nil {
+		// If the pod is not found, we should create it.
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+		// All other errors are actual errors.
+	} else {
+		found = true
+	}
+
+	if found {
 		log.Info("Found Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
 		// TODO - Do we enfore (change) spec if a pod exists ?
 		// r.Patch(ctx, pod, client.Apply, client.ForceOwnership, client.FieldOwner("sandbox-controller"))
```

Comment:
Either way, we probably want to make the choice explicit by logging, event message or something else eventually.